src/main/java/appeng/container/AEBaseContainer.java

```diff
@@ -418,18 +402,21 @@ public ItemStack transferStackInSlot( final EntityPlayer p, final int idx )
 			 */
 			if( clickSlot.isPlayerSide() )
 			{
-				tis = this.shiftStoreItem( tis );
+				tis = this.transferStackToContainer( tis );
 
-				// target slots in the container...
-				for( final Object inventorySlot : this.inventorySlots )
+				if( !tis.isEmpty() )
 				{
-					final AppEngSlot cs = (AppEngSlot) inventorySlot;
-
-					if( !( cs.isPlayerSide() ) && !( cs instanceof SlotFake ) && !( cs instanceof SlotCraftingMatrix ) )
+					// target slots in the container...
+					for( final Object inventorySlot : this.inventorySlots )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)